core/drivers/crypto/caam/utils/utils_dmaobj.c

```diff
@@ -0,0 +1,1395 @@
+		     input, output, *length, off, align);
+
+	if (!input || !output || !length || !input->priv || !output->priv ||
+	    !*length) {
+		ret = TEE_ERROR_BAD_PARAMETERS;
+		goto end;
+	}
+
+	/*
+	 * First build the input SGT/Buffer
+	 */
+	ret = caam_dmaobj_sgtbuf_build(input, length, off, align);
+	if (ret)
+		goto end;
+
+	/*
+	 * Next build the output SGT/Buffer.
+	 * If returned length is not same as input, redo the input
+	 * SGT/Buffer with the same length as the output.
+	 */
+	len = *length;
+	ret = caam_dmaobj_sgtbuf_build(output, &len, off, *length);
+	if (ret)
+		goto end;
+
+	if (len != *length) {
+		DMAOBJ_TRACE("Retry In %zu bytes vs Out %zu bytes", *length,
+			     len);
+
+		/* Redo the input with the output length */
+		*length = len;
+		ret = caam_dmaobj_sgtbuf_build(input, length, off, len);
+		if (!ret && *length != len) {
+			DMAOBJ_TRACE("Error In %zu bytes vs Out %zu bytes",
+				     *length, len);
+			ret = TEE_ERROR_OUT_OF_MEMORY;
+		}
+	}
+
+end:
+	DMAOBJ_TRACE("Input/Output SGTBUF returns 0x%" PRIx32, ret);
+
+	return ret;
+}
+
+TEE_Result caam_dmaobj_sgtbuf_build(struct caamdmaobj *obj, size_t *length,
+				    size_t off, size_t align)
+{
+	TEE_Result ret = TEE_ERROR_GENERIC;
+	enum caam_status retstatus = CAAM_FAILURE;
+	struct priv_dmaobj *priv = NULL;
+	struct dmaentry *entry = NULL;
+	struct dmaentry *start_entry = NULL;
+	size_t max_length = 0;
+	size_t acc_length = 0;
+	size_t offset = off;
+	unsigned int idx = 0;
+	unsigned int nb_sgt = 0;
+
+	DMAOBJ_TRACE("obj=%p of %zu bytes (offset=%zu) - align %zu", obj,
+		     *length, off, align);
+
+	if (!obj || !obj->priv || !length || !*length) {
+		ret = TEE_ERROR_BAD_PARAMETERS;
+		goto end;
+	}
+
+	priv = obj->priv;
+
+	max_length = *length;
+	if (priv->dmabuf.allocated && max_length > priv->dmabuf.allocated &&
+	    priv->dmabuf.allocated > align)
+		max_length = ROUNDDOWN(priv->dmabuf.allocated, align);
+
+	DMAOBJ_TRACE("Prepare SGT/Buffer to do %zu of %zu", max_length,
+		     *length);
+
+	/* Find the first DMA buffer to start with */
+	TAILQ_FOREACH(entry, &priv->list, next)
+	{
+		if (offset < entry->origbuf.length)
+			break;
+
+		offset -= entry->origbuf.length;
+	}
+
+	if (!entry) {
+		DMAOBJ_TRACE("There is no DMA Object available");
+		ret = TEE_ERROR_GENERIC;
+		goto end;
+	}
+
+	start_entry = entry;
+	DMAOBJ_TRACE("Start with %p data %p offset %zu", start_entry,
+		     start_entry->origbuf.data, offset);
+
+	acc_length = entry->origbuf.length - offset;
+	nb_sgt = 1;
+
+	/* Calculate the number of SGT entry */
+	for (entry = TAILQ_NEXT(entry, next); entry && acc_length < max_length;
+	     entry = TAILQ_NEXT(entry, next)) {
+		acc_length += entry->origbuf.length;
+		nb_sgt++;
+	}
+
+	DMAOBJ_TRACE("%d of %d SGT/Buffer entries to handle", nb_sgt,
+		     priv->nb_sgtbuf);
+	if (priv->nb_sgtbuf < nb_sgt) {
+		if (priv->nb_sgtbuf) {
+			obj->sgtbuf.number = priv->nb_sgtbuf;
+			obj->sgtbuf.sgt_type =
```

Comment:
This is the same as `obj->sgtbuf.sgt_type = (priv->nb_sgtbuf > 1);`, the `()` are redundant but I think it makes it a bit easier to read.

Reply:
Ok